### ISSUE TITLE
mfsmount hangs after a while.

### ISSUE THREAD

**Issue by k4SsKInGlUJGlnFp** [moosefs/moosefs#64]
Hello, 

I am trying to use MooseFS as the backend for Nginx + PHP-FPM serving websites. I've found that after a short while, the systems running php-fpm lock up when trying to access the fuse mount. Before the system locks, the following errors show up in /var/log/messages. 

`mfsmount[54885]: sysctl(kern.proc.filedesc) error: EBUSY (Device or resource busy)`
`mfsmount[54885]: readworker: can't set TCP_NODELAY: ECONNRESET (Connection reset by peer)`

Those are followed by lots of these,
`kernel: vm_fault: pager read error, pid 19476 (php-fpm)`

Eventually, the php-fpm process enter a hung state and trying to access the mfsmount hangs as well. Even a simple 'ls' command fails. Even though the fuse mount and php-fpm processes are hung, the actual system stays up and is responsive so long as I do not attempt anything that touches the mfsmount. 

The setup is 1x Master, 2x chunk servers.  
1x Master = moosefs3-master-3.0.94
2x Chunk = moosefs3-chunkserver-3.0.94
PHP-FPM = PHP 7.0.21
Nginx = 1.12.1
Test software = ownCloud
FreeBSD = 11.0

I'm not sure if this is an issue in FreeBSD, Fuse, Moosefs, or simply bad configuration on my part? The MooseFS configuration is the defaults provided by the FreeBSD port. 

Any advice appreciated. I can change settings and retest as needed.

**Comment by k4SsKInGlUJGlnFp** [comment]
The above errors seem to be defined here, https://www.freebsd.org/cgi/man.cgi?query=errno&apropos=0&sektion=2&manpath=FreeBSD+10.0-RELEASE&arch=default&format=html

I'm not sure how they are related to Moosefs other than being triggered by it.

**EBUSY** Device busy.  An attempt to use a	system resource	which was in
	     use at the	time in	a manner which would have conflicted with the
	     request.


**ECONNRESET** Connection reset by peer.  A	connection was forcibly	closed
	     by	a peer.	 This normally results from a loss of the connection
	     on	the remote socket due to a timeout or a	reboot.

None of the systems were shutdown or rebooted, so I'm not sure why this was happening.

**Comment by xandrus** [comment]
Hi.
First of all please update to MosoeFS 3.0.96.
Version 3.0.94 has a bug with dirattrcache which influence on MooseFS client performance and can cause all this errors.

**Comment by k4SsKInGlUJGlnFp** [comment]
Hi, 

Good to know. I'll update as soon as the ports tree has it and report back. 

Thank you

**Comment by oxide94** [comment]
Hi @k4SsKInGlUJGlnFp,

Today I [created a bug on FreeBSD Bugzilla](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=221183) with update to version 3.0.97. [It has just been merged](https://svnweb.freebsd.org/ports?view=revision&revision=447136) with official FreeBSD ports tree.

Please update your instance to MooseFS 3.0.97.

By the way - you can always use official ports available [here](https://github.com/moosefs/moosefs-freebsd-ports) or FreeBSD packages available [here](https://moosefs.com/download/freebsd.html).

Best regards,
Peter / MooseFS Team

**Comment by k4SsKInGlUJGlnFp** [comment]
Hello, 

I updated to MooseFS 3.0.97 and retested. I am seeing the same errors. Here is the output from /var/log/messages.

> Aug  2 22:32:29 grid2 pkg-static: moosefs3-client-3.0.97 installed
Aug  2 22:48:11 grid2 kernel: fuse-freebsd: version 0.4.4, FUSE ABI 7.8
Aug  2 22:48:12 grid2 mfsmount[11219]: monotonic clock function: clock_gettime
Aug  2 22:48:12 grid2 mfsmount[11219]: monotonic clock speed: 200723 ops / 10 mili seconds
Aug  2 22:48:13 grid2 mfsmount[11219]: my st_dev: 3976396564
Aug  2 23:44:06 grid2 kernel: vm_fault: pager read error, pid 67272 (php-fpm)
Aug  3 00:04:12 grid2 kernel: vm_fault: pager read error, pid 58995 (php-fpm)
Aug  3 01:25:15 grid2 mfsmount[11219]: sysctl(kern.proc.filedesc) error: EBUSY (Device or resource busy)
Aug  3 04:02:33 grid2 kernel: vm_fault: pager read error, pid 8739 (php-fpm)

It takes a few hours from the start of the errors to the point where the mfsmount hangs. For this test, the software accessing the mount point was PHP-FPM and Nginx running 5 different websites.

Any ideas, or suggestions on how to debug or resolve this?

**Comment by k4SsKInGlUJGlnFp** [comment]
Any hints on how to debug/fix this? This has been a repeatable and easy to reproduce issue on my test environment.

**Comment by xandrus** [comment]
Hi.
I have additional question. 
Do you have any older MooseFS mounts like 2.0.x in your cluster or all clients are in 3.0.97 version?

**Comment by xandrus** [comment]
Hi.
By the way you can try to mount MooseFS client with cache mode set to AUTO:

`mfsmount -o mfscachemode=AUTO /mnt/mfs`

By default on FreeBSD cache mode is set to DIRECT because it is much more efficient, but it is possible that in DIRECT mode "mmap" doesn't work (see #66). Nginx might use mmap to improve performance and maybe this causes this issue. So please check if it resolves this problem. In the meantime we will try to reproduce it on our testing environment.

**Comment by k4SsKInGlUJGlnFp** [comment]
Hi Xandrus, 

All the clients are 3.0.97. I'll try using mfscachemode=AUTO today.

**Comment by k4SsKInGlUJGlnFp** [comment]
Hello, 

Unfortunately, that did not make a difference. I've tried it twice, and I have the same errors and eventually mfsmount hang up. 

Once the mfsmount hangs, the server has to be rebooted to remove it. Trying fstat, or lsof on the mount point causes those programs to hang as well. 

I've noticed I can successfully run MooseFS with a low traffic site. The lock up happens only when I run it with a site receiving consistent traffic. The higher the concurrent requests, the faster the mfsmount hangs.

**Comment by xandrus** [comment]
Thank you for all this information.
We ware trying to reproduce this scenario in our development environment but without success.

Would you be so kind and send us MooseFS oplog during consistent trafic from site, so we will be able to see what kind of sequence causing this freeze. You can easily do it like this:

`cat /mnt/mfs/.oplog | tee ~/mfs-oplog.log`

And after freeze hit ctrl+c

**Comment by k4SsKInGlUJGlnFp** [comment]
Hello, 

I'm sorry, but I'm not seeing an '.oplog' anywhere on any of the systems. I've checked the Master, the chunk servers, and the clients. Also, there is no /etc/mfs/ folder. There is /usr/local/etc/mfs, and /var/mfs, but neither have an '.oplog' file. 

Is there someplace else to look, or a configuration I need to set to generate that log?

**Comment by oxide94** [comment]
Hi @k4SsKInGlUJGlnFp,

`.oplog` is an internal, hidden MooseFS-specific file. It is not visible while running `ls -a`. It is available in root path of any MooseFS mountpoint, so e.g. if you have MooseFS mounted in `/mnt/mfs`, path to `.oplog` is `/mnt/mfs/.oplog`, as @xandrus mentioned above. Just run the command provided by him.

Thanks,
Peter / MooseFS Team

**Comment by k4SsKInGlUJGlnFp** [comment]
Ahh, got it. 

I'm testing it now. If this follows normal patterns, it will be 8 or 9 hours before it locks up. 

Thank you

**Comment by xandrus** [comment]
Hi.
I would like to inform you that we have just configured similar environment and we are executing some load tests.
By the way. Would you be so kind and contact us directly at bugs@moosefs.com so you will be able to send us large files with debugs information.

**Comment by k4SsKInGlUJGlnFp** [comment]
An email with additional details and the complete log has been sent to bugs@moosefs.com

Thank you

**Comment by butonic** [comment]
Was this resolved?

**Comment by k4SsKInGlUJGlnFp** [comment]
Hello, 

I never found a certain cause. During the testing, MooseFS was run in jails on the same physical systems as the application servers (php-fpm, nginx). I think the issue may have been due to resource starvation, or network communication issues. (Ex, firewall interference)

MooseFS has since been moved to dedicated hardware, on a private internal network with jumbo frames. I have not been able to replicate the issue since. 

In short, I think our testing environment was simply not up to task.

This does not seem to be a MooseFS issue. Lets close this out.
